docs/5-references/2-configuration.md
# Outpost Configuration

Global configurations are provided through env variables or a YAML file. ConfigMap can be used if deploying with Kubernetes.

| Variable                                         | Default                                 | Required                                                           |
| ------------------------------------------------ | --------------------------------------- | ------------------------------------------------------------------ |
| `SERVICE`                                        | `nil`                                   | No                                                                 |
| `CONFIG`                                         | `nil`                                   | No                                                                 |
|                                                  |                                         |                                                                    |
| `ORGANIZATION_NAME`                              | `default`                               | Yes                                                                |
|                                                  |                                         |                                                                    |
| `API_KEY`                                        | `nil`                                   | Yes                                                                |
| `API_PORT`                                       | `3333`                                  | Yes                                                                |
| `API_JWT_SECRET`                                 | `nil`                                   | Only for using JWT Auth                                            |
|                                                  |                                         |                                                                    |
| `AES_ENCRYPTION_SECRET`                          | `nil`                                   | Yes                                                                |
|                                                  |                                         |                                                                    |
| `TOPICS`                                         | `''`                                    | No                                                                 |
|                                                  |                                         |                                                                    |
| `REDIS_PORT`                                     | `6379`                                  | Yes                                                                |
| `REDIS_HOST`                                     | `127.0.0.1`                             | Yes                                                                |
| `REDIS_PASSWORD`                                 | `nil`                                   | Yes                                                                |
| `REDIS_DATABASE`                                 | `0`                                     | Yes                                                                |
| `CLICKHOUSE_USER`                                | `nil`                                   | Yes                                                                |
| `CLICKHOUSE_DATABASE`                            | `outpost`                               | Yes                                                                |
| `CLICKHOUSE_HOST`                                | `nil`                                   | Yes                                                                |
| `CLICKHOUSE_PASSWORD`                            | `nil`                                   | Yes                                                                |
|                                                  |                                         |                                                                    |
| `RABBITMQ_SERVER_URL`                            | `nil`                                   | No                                                                 |
| `RABBITMQ_EXCHANGE`                              | `outpost`                               | No                                                                 |
| `RABBITMQ_DELIVERY_QUEUE`                        | `outpost-delivery`                      | No                                                                 |
| `RABBITMQ_LOG_QUEUE`                             | `outpost-log`                           | No                                                                 |
|                                                  |                                         |                                                                    |
| `AWS_SQS_REGION`                                 | `nil`                                   | No                                                                 |
| `AWS_SQS_ACCESS_KEY_ID`                          | `nil`                                   | No                                                                 |
| `AWS_SQS_SECRET_ACCESS_KEY`                      | `nil`                                   | No                                                                 |
| `AWS_SQS_DELIVERY_QUEUE`                         | `outpost-delivery`                      | No                                                                 |
| `AWS_SQS_LOG_QUEUE`                              | `outpost-log`                           | No                                                                 |
|                                                  |                                         |                                                                    |
| `GCP_PUBSUB_SERVICE_ACCOUNT_CREDS`               | `nil`                                   | No                                                                 |
| `GCP_PUBSUB_DELIVERY_TOPIC`                      | `outpost-delivery`                      | No                                                                 |
| `GCP_PUBSUB_LOG_TOPIC`                           | `outpost-log`                           | No                                                                 |
|                                                  |                                         |                                                                    |
| `PUBLISH_RABBITMQ_SERVER_URL`                    | `nil`                                   | No                                                                 |
| `PUBLISH_RABBITMQ_QUEUE`                         | `nil`                                   | No                                                                 |
| `PUBLISH_AWS_REGION`                             | `nil`                                   | No                                                                 |
| `PUBLISH_AWS_SQS_ACCESS_KEY_ID`                  | `nil`                                   | No                                                                 |
| `PUBLISH_AWS_SQS_SECRET_ACCESS_KEY`              | `nil`                                   | No                                                                 |
| `PUBLISH_AWS_SQS_QUEUE`                          | `nil`                                   | No                                                                 |
| `PUBLISH_GCP_PUBSUB_SERVICE_ACCOUNT_CREDS`       | `nil`                                   | No                                                                 |
| `PUBLISH_GCP_PUBSUB_SUBSCRIPTION`                | `nil`                                   | No                                                                 |
|                                                  |                                         |                                                                    |
| `PUBLISH_MAX_CONCURRENCY`                        | `10`                                    | No                                                                 |
| `DELIVERY_MAX_CONCURRENCY`                       | `10`                                    | Yes                                                                |
| `LOG_MAX_CONCURRENCY`                            | `10`                                    | Yes                                                                |
| `LOG_RETRY_LIMIT`                                | `5`                                     | Yes                                                                |
|                                                  |                                         |                                                                    |
| `RETRY_INTERVAL_SECONDS`                         | `30`                                    | Yes                                                                |
| `MAX_RETRY_LIMIT`                                | `10`                                    | Yes                                                                |
| `DELIVERY_TIMEOUT_SECONDS`                       | `5`                                     | Yes                                                                |
| `HTTP_USER_AGENT`                                | `Outpost 1.0`                           | Yes                                                                |
| `MAX_EVENT_SIZE_KB`                              | `256`                                   | Yes                                                                |
| `MAX_DESTINATIONS_PER_TENANT`                    | `20`                                    | Yes                                                                |
| `LOG_BATCH_SIZE`                                 | `1000`                                  | Yes                                                                |
| `LOG_BATCH_THRESHOLD_SECONDS`                    | `10`                                    | Yes                                                                |
|                                                  |                                         |                                                                    |
| `DESTINATIONS_WEBHOOK_HEADER_PREFIX`             | `x-`                                    | No                                                                 |
| `DESTINATIONS_WEBHOOK_DISABLE_EVENT_ID_HEADER`   | `false`                                 | No                                                                 |
| `DESTINATIONS_WEBHOOK_DISABLE_SIGNATURE_HEADER`  | `false`                                 | No                                                                 |
| `DESTINATIONS_WEBHOOK_DISABLE_TIMESTAMP_HEADER`  | `false`                                 | No                                                                 |
| `DESTINATIONS_WEBHOOK_DISABLE_TOPIC_HEADER`      | `false`                                 | No                                                                 |
| `DESTINATIONS_WEBHOOK_SIGNATURE_VALUE_TEMPLATE`  | `{{.Timestamp.Unix}}.{{.Body}}`         | No                                                                 |
| `DESTINATIONS_WEBHOOK_SIGNATURE_HEADER_TEMPLATE` | `t={{.Timestamp.Unix}},v0={{.Signatures | join ","}}`                                                        |
| `DESTINATIONS_WEBHOOK_SIGNATURE_ENCODING`        | `hex`                                   | No                                                                 |
| `DESTINATIONS_WEBHOOK_SIGNATURE_ALGORITHM`       | `hmac-sha256`                           | No                                                                 |
|                                                  |                                         |                                                                    |
| `ALERT_CALLBACK_URL`                             | `nil`                                   | No                                                                 |
| `ALERT_DEBOUNCING_INTERVAL_SECOND`               | `3600`                                  | No                                                                 |
| `ALERT_CONSECUTIVE_FAILURE_COUNT`                | `20`                                    | No                                                                 |
| `ALERT_FAILURE_WINDOW_SECOND`                    | `60`                                    | No                                                                 |
| `ALERT_FAILURE_RATE`                             | `0.1`                                   | No                                                                 |
| `ALERT_AUTO_DISABLE_DESTINATION`                 | `true`                                  | No                                                                 |
|                                                  |                                         |                                                                    |
| `PORTAL_REFERER_URL`                             | `nil`                                   | Yes                                                                |
| `PORTAL_FAVICON_URL`                             | `nil`                                   | No                                                                 |
| `PORTAL_LOGO`                                    | `nil`                                   | No                                                                 |
| `PORTAL_FORCE_THEME`                             | `nil`                                   | No                                                                 |
| `PORTAL_ACCENT_COLOR`                            | `nil`                                   | No                                                                 |
| `PORTAL_BRAND_COLOR`                             | `nil`                                   | No                                                                 |
| `PORTAL_OUTPOST_BRANDING`                        | `true`                                  | No                                                                 |
|                                                  |                                         |                                                                    |
| `DISABLE_TELEMETRY`                              | `false`                                 | Yes                                                                |
| `LOG_LEVEL`                                      | `info`                                  | Yes                                                                |
| `AUDIT_LOG`                                      | `true`                                  | Yes                                                                |
|                                                  |                                         |                                                                    |
| `OTEL_SERVICE_NAME`                              | `nil`                                   | No                                                                 |
| `OTEL_*`                                         | `nil`                                   | https://opentelemetry.io/docs/languages/sdk-configuration/general/ |
| `DESTINATIONS_METADATA_PATH`                     | `config/outpost/destinations`           | No                                                                 |


internal/config/config.go
package config

import (
	"errors"
	"log"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"

	"github.com/hookdeck/outpost/internal/clickhouse"
	"github.com/hookdeck/outpost/internal/mqs"
	"github.com/hookdeck/outpost/internal/otel"
	"github.com/hookdeck/outpost/internal/redis"
	"github.com/joho/godotenv"
	v "github.com/spf13/viper"
)

const (
	Namespace = "Outpost"
)

type Config struct {
	Service          ServiceType
	Port             int
	Hostname         string
	APIKey           string
	JWTSecret        string
	EncryptionSecret string
	PortalProxyURL   string
	Topics           []string

	Redis                           *redis.RedisConfig
	ClickHouse                      *clickhouse.ClickHouseConfig
	OpenTelemetry                   *otel.OpenTelemetryConfig
	PublishQueueConfig              *mqs.QueueConfig
	DeliveryQueueConfig             *mqs.QueueConfig
	LogQueueConfig                  *mqs.QueueConfig
	PublishMaxConcurrency           int
	DeliveryMaxConcurrency          int
	LogMaxConcurrency               int
	RetryIntervalSeconds            int
	RetryMaxCount                   int
	LogBatcherDelayThresholdSeconds int
	LogBatcherItemCountThreshold    int
	DestinationMetadataPath         string

	DisableTelemetry bool

	DestinationWebhookHeaderPrefix string

	// Portal config
	PortalRefererURL             string
	PortalFaviconURL             string
	PortalLogo                   string
	PortalOrgName                string
	PortalForceTheme             string
	PortalDisableOutpostBranding bool
}

var defaultConfig = map[string]any{
	"PORT":                                3333,
	"REDIS_HOST":                          "127.0.0.1",
	"REDIS_PORT":                          6379,
	"REDIS_PASSWORD":                      "",
	"REDIS_DATABASE":                      0,
	"DELIVERY_RABBITMQ_EXCHANGE":          "outpost",
	"DELIVERY_RABBITMQ_QUEUE":             "outpost.delivery",
	"LOG_RABBITMQ_EXCHANGE":               "outpost_logs",
	"LOG_RABBITMQ_QUEUE":                  "outpost_logs.log",
	"PUBLISHMQ_MAX_CONCURRENCY":           1,
	"DELIVERYMQ_MAX_CONCURRENCY":          1,
	"LOGMQ_MAX_CONCURRENCY":               1,
	"RETRY_INTERVAL_SECONDS":              30,
	"MAX_RETRY_COUNT":                     10,
	"LOG_BATCHER_DELAY_THRESHOLD_SECONDS": 5,
	"LOG_BATCHER_ITEM_COUNT_THRESHOLD":    100,
	"DESTINATION_METADATA_PATH":           "config/outpost/destinations",
	"DESTINATION_WEBHOOK_HEADER_PREFIX":   "x-outpost-",
}

var (
	ErrMismatchedServiceType = errors.New("service type mismatch")
)

func Parse(flags Flags) (*Config, error) {
	var err error

	// Use a local Viper instance to avoid leaking configuration to global scope
	viper := v.New()

	hostname, err := os.Hostname()
	if err != nil {
		return nil, err
	}

	// Load .env file to environment variables
	err = godotenv.Load()
	if err != nil {
		// Ignore error if file does not exist
	}

	// Set default config values
	for key, value := range defaultConfig {
		viper.SetDefault(key, value)
	}

	// Parse custom config file if provided
	if flags.Config != "" {
		viper.SetConfigFile(flags.Config)
		if err := viper.ReadInConfig(); err != nil {
			return nil, err
		}
	}

	// Bind environemnt variable to viper
	viper.AutomaticEnv()

	// Parse service type from flag & env
	service, err := parseService(viper, flags)
	if err != nil {
		return nil, err
	}

	openTelemetry, err := parseOpenTelemetryConfig(viper)
	if err != nil {
		return nil, err
	}

	var clickHouseConfig *clickhouse.ClickHouseConfig
	if viper.GetString("CLICKHOUSE_ADDR") != "" {
		clickHouseConfig = &clickhouse.ClickHouseConfig{
			Addr:     viper.GetString("CLICKHOUSE_ADDR"),
			Username: viper.GetString("CLICKHOUSE_USERNAME"),
			Password: viper.GetString("CLICKHOUSE_PASSWORD"),
			Database: viper.GetString("CLICKHOUSE_DATABASE"),
		}
	}

	topics, err := parseTopics(viper)
	if err != nil {
		return nil, err
	}

	// MQs
	publishQueueConfig, err := mqs.ParseQueueConfig(viper, "PUBLISH")
	if err != nil {
		return nil, err
	}
	deliveryQueueConfig, err := mqs.ParseQueueConfig(viper, "DELIVERY")
	if err != nil {
		return nil, err
	}
	deliveryQueueConfig.Policy.RetryLimit = viper.GetInt("DELIVERYMQ_RETRY_LIMIT")
	logQueueConfig, err := mqs.ParseQueueConfig(viper, "LOG")
	if err != nil {
		return nil, err
	}
	logQueueConfig.Policy.RetryLimit = viper.GetInt("LOGMQ_RETRY_LIMIT")

	portalProxyURL := viper.GetString("PORTAL_PROXY_URL")
	if portalProxyURL != "" {
		if _, err := url.Parse(portalProxyURL); err != nil {
			return nil, err
		}
	}

	// Initialize config values
	config := &Config{
		Hostname:         hostname,
		Service:          *service,
		Port:             getPort(viper),
		APIKey:           viper.GetString("API_KEY"),
		JWTSecret:        viper.GetString("JWT_SECRET"),
		EncryptionSecret: viper.GetString("ENCRYPTION_SECRET"),
		PortalProxyURL:   portalProxyURL,
		Topics:           topics,
		Redis: &redis.RedisConfig{
			Host:     viper.GetString("REDIS_HOST"),
			Port:     mustInt(viper, "REDIS_PORT"),
			Password: viper.GetString("REDIS_PASSWORD"),
			Database: mustInt(viper, "REDIS_DATABASE"),
		},
		ClickHouse:                      clickHouseConfig,
		OpenTelemetry:                   openTelemetry,
		PublishQueueConfig:              publishQueueConfig,
		DeliveryQueueConfig:             deliveryQueueConfig,
		LogQueueConfig:                  logQueueConfig,
		PublishMaxConcurrency:           mustInt(viper, "PUBLISHMQ_MAX_CONCURRENCY"),
		DeliveryMaxConcurrency:          mustInt(viper, "DELIVERYMQ_MAX_CONCURRENCY"),
		LogMaxConcurrency:               mustInt(viper, "LOGMQ_MAX_CONCURRENCY"),
		RetryIntervalSeconds:            mustInt(viper, "RETRY_INTERVAL_SECONDS"),
		RetryMaxCount:                   mustInt(viper, "MAX_RETRY_COUNT"),
		LogBatcherDelayThresholdSeconds: mustInt(viper, "LOG_BATCHER_DELAY_THRESHOLD_SECONDS"),
		LogBatcherItemCountThreshold:    mustInt(viper, "LOG_BATCHER_ITEM_COUNT_THRESHOLD"),
		DestinationMetadataPath:         viper.GetString("DESTINATION_METADATA_PATH"),

		DisableTelemetry: viper.GetBool("DISABLE_TELEMETRY"),

		DestinationWebhookHeaderPrefix: viper.GetString("DESTINATION_WEBHOOK_HEADER_PREFIX"),

		// Portal config
		PortalRefererURL:             viper.GetString("PORTAL_REFERER_URL"),
		PortalFaviconURL:             viper.GetString("PORTAL_FAVICON_URL"),
		PortalLogo:                   viper.GetString("PORTAL_LOGO"),
		PortalOrgName:                viper.GetString("PORTAL_ORGANIZATION_NAME"),
		PortalForceTheme:             viper.GetString("PORTAL_FORCE_THEME"),
		PortalDisableOutpostBranding: viper.GetBool("PORTAL_DISABLE_OUTPOST_BRANDING"),
	}

	return config, nil
}

func mustInt(viper *v.Viper, configName string) int {
	i, err := strconv.Atoi(viper.GetString(configName))
	if err != nil {
		log.Fatalf("%s = '%s' is not int", configName, viper.GetString(configName))
	}
	return i
}

func getPort(viper *v.Viper) int {
	port := mustInt(viper, "PORT")
	if viper.GetString("API_PORT") != "" {
		apiPort, err := strconv.Atoi(viper.GetString("API_PORT"))
		if err == nil {
			port = apiPort
		}
	}
	return port
}

func parseService(viper *v.Viper, flags Flags) (*ServiceType, error) {
	serviceString := flags.Service
	if viper.GetString("SERVICE") != "" {
		if serviceString != "" && serviceString != viper.GetString("SERVICE") {
			return nil, ErrMismatchedServiceType
		}
		serviceString = viper.GetString("SERVICE")
	}
	service, err := ServiceTypeFromString(serviceString)
	if err != nil {
		return nil, err
	}
	return &service, nil
}

func parseTopics(viper *v.Viper) ([]string, error) {
	topicStr := viper.GetString("TOPICS")
	if topicStr == "" {
		return nil, makeEmptyErr("TOPICS")
	}
	topics := strings.Split(topicStr, ",")
	for i, topic := range topics {
		topics[i] = strings.TrimSpace(topic)
	}
	sort.Strings(topics)
	return topics, nil
}

func makeEmptyErr(key string) error {
	return errors.New(key + " is empty")
}
